report/41_Lessons-Evolution-&-Refactoring.md

```diff
@@ -1 +1,101 @@
+
+The team knew that choosing scalable techonologies early was important.
+If a choice to switch from a monolithic server to a distributed system comes too late, then the changes that will have to be made could become quite big, resulting in a lot of technical debt.
+
+This was why dockerization came early for the project, with separation between components for easier clusterization.
+The team didn't quite get Docker Swarm to work, but because these considerations were done early, Docker Swarm was actually obtainable, instead of being buried in debt from not having a RESTful API, for example.
+
+#### Graphs are motivational
+
+After the simulator had been live for a few weeks, the statistics showed that this project lacked behind the other participating projects by quite a big margin.
+Having this as a motivator to improve was vital to finding some of the bugs in the system.
+
+First, it was discovered that the project lacked behind because response times were highly influential for the graph trend lines.
+It was then considered whether HTTPS and Traefik routing were the thing slowing down the simulator, but removing those things only made very slight improvements, so they were re-enabled.
+The actual reason turned out to be the server location, as will be described in [TODO: 4.2 Operations](42_Lessons-Operation.md).
+
+Without the graph, the statistics, and the competitive nature of the team as motivators, these things would never have been discovered.
+Replication of this feeling would be difficult, as there rarely are 10 teams competing to implement the best and fastest project when looking at the job market.
+Gamification of collected metrics might be able to lead teams to improve these things, for example through having a big screen with several desired trend lines/thresholds along with actual statistics.
+The more prominent the gamification, the more this competitive nature could drive the team to do better.
+
+#### Proper communication
+
+A problem that the team never got to improve on was communication, especially of the asynchronous kind.
+When co-located, the team managed to get lots of work done and reviewed.
+At the times between co-located sessions and after the pandemic hit the world, the reviews were slow and impractical.
+
+A possible reason could be that reminding people to review is easier and more direct when working next to each other.
+Another could be that the other courses and work got in the way.
+
+A third possibility could be that there were no channels for politely nudging towards a review.
+The team tried to alleviate this with the `#review` channel on Discord, where anyone could push for a review.
+This helped in some ways, as Discord is a more prominently used communication tool among the developers.
+
+Though the team hasn't found better solutions than just being more demanding towards each other, it acknowledges that this is a problem that needs improving somehow down the line.
```

Comment:
I wrote about something similiar in 4.3, it might be worth making them fitter better together.